troll-o-matic.py

```diff
@@ -273,27 +284,53 @@ def process_changes(self, prefix, changes):
 
       gerrit_patch = rev.get_commit_from_remote('cros', cur_rev.ref)
 
-      upstream_shas = rev.get_cherry_pick_shas_from_patch(gerrit_patch)
-      if not upstream_shas:
-        self.handle_missing_hash_review(c, prefix)
-        continue
+      if prefix == 'FROMLIST':
+        upstream_patchworks = rev.get_am_from_from_patch(gerrit_patch)
+        if not upstream_patchworks:
+          self.handle_missing_am_review(c, prefix)
+          continue
 
-      upstream_patch = None
-      upstream_sha = None
-      for s in reversed(upstream_shas):
-        try:
-          upstream_patch = rev.get_commit_from_sha(s)
-          upstream_sha = s
-          break
-        except:
+        upstream_patch = None
+        for u in reversed(upstream_patchworks):
+          try:
+            upstream_patch = rev.get_commit_from_patchwork(u)
+            break
+          except:
+            continue
+
+        if not upstream_patch:
+          self.print_error(
+            'ERROR: patch missing from patchwork, or patchwork host '
+            'not whitelisted for {} ({})\n'.format(
+                                    c, upstream_patchworks))
+          self.blacklist.append(c)
+          continue
+      else:
+        upstream_shas = rev.get_cherry_pick_shas_from_patch(gerrit_patch)
+        if not upstream_shas:
+          self.handle_missing_hash_review(c, prefix)
           continue
-      if not upstream_patch:
-        self.print_error('ERROR: SHA missing from git for {} ({})\n'.format(
+
+        upstream_patch = None
+        upstream_sha = None
+        for s in reversed(upstream_shas):
+          try:
+            upstream_patch = rev.get_commit_from_sha(s)
+            upstream_sha = s
+            break
+          except:
+            continue
+
+        if not upstream_patch:
+          self.print_error('ERROR: SHA missing from git for {} ({})\n'.format(
                                     c, upstream_shas))
-        self.blacklist.append(c)
-        continue
+          self.blacklist.append(c)
+          continue
 
-      fixes_ref = rev.find_fixes_reference(upstream_sha)
+      if prefix != 'FROMLIST':
+        fixes_ref = rev.find_fixes_reference(upstream_sha)
+      else:
+        fixes_ref = None
 
       result = rev.compare_diffs(upstream_patch, gerrit_patch)
```

Comment:
I think this is going to be the tricky part, and this is why I haven't done FROMLIST.

git's diff algorithm differed enough between versions that I was getting a bunch of false negatives when grabbing the patch from gerrit or the list. That's why I need to fetch and apply the patches from gerrit to my local tree before comparing the diffs.

I fear that this will produce a lot of noise. We might only want to comment on FROMLIST if the backport is clean, and ignore the ones that differ.

Reply:
Oh, I see. Maybe we could try to apply to patch locally first, as well? I wonder how things like 0day deal with dependencies, though...

Being able to tag clean and obvious backports would already be good, IMHO.